### PR TITLE
Define and activate kvm 12.1.0 release

### PR DESCRIPTION
**Pull request by sslavic**
Towards https://github.com/giantswarm/giantswarm/issues/12391

Includes:
- https://github.com/giantswarm/giantswarm/issues/12188
- https://github.com/giantswarm/giantswarm/issues/12064

### CODE REVIEW

**Comment by sslavic** [comment]
Tested on geckon:
- 12.1.0 cluster creation
- upgrade from 11.3.2

**Comment by sslavic** [comment]
@giantswarm/chapter-se unlike for Azure, unfortunately changing nginx Service ownership in KVM case includes brief downtime.

Existing nginx NodePort Service has to be manually deleted before cluster upgrade starts. As soon as nginx NodePort Service gets deleted please trigger the upgrade. Upgrade includes nginx App release upgrade which will install a replacement nginx NodePort Service with same name and same node ports as one that was managed by kvm-operator. Typically all included app upgrades happen before master goes down, so downtime is relatively short ~30sec (depending on how old platform release cluster is being upgraded from, newer it is shorter is the app upgrade and downtime).

To minimize this unwanted side-effect we need your help. Once release is out please arrange cluster upgrade maintenance window with customers on KVM, and assist with the cluster upgrade itself.

**Comment by sslavic** [comment]
Did one more round of tests on geckon to verify 12.1.0 cluster creation still works after switching from dev to operator releases. Deployed also two nginx apps using this config override for second nginx:

```
controller:
  ingressClass: "nginx-internal"
  service:
    enabled: true
    public: false
    subdomain: "ingress-internal"
    type: NodePort
    nodePorts:
      http: 31010
      https: 31011
```

File: kvm/v12.1.0/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+## :zap:  Giant Swarm Release 12.1.0 for KVM :zap:
+
+In this release:
```
Comment (cokiengchiara) [inline kvm/v12.1.0/README.md:3]:
```suggestion
**This release includes two significant improvements to NGINX Ingress Controller. It also adds support for image registry mirrors, improving reliability.**
```

Just to give a top level summary.

File: kvm/v12.1.0/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+- Multiple NGINX Ingress Controllers per tenant cluster are supported.
```
Comment (cokiengchiara) [inline kvm/v12.1.0/README.md:5]:
```suggestion
**The two NGINX Ingress Controller improvements:**
* First, multiple NGINX Ingress Controllers per tenant cluster are now supported, enabling separation of internal vs external traffic, dev vs prod, and so on.
* Second, management of NGINX IC NodePort Service is moved from kvm-operator to NGINX IC App itself. This enables configurability of external traffic policy and lays the foundation for making NGINX IC App optional and not pre-installed in a future KVM platform release.
```

File: kvm/v12.1.0/README.md (hunk 1 above)
Comment (cokiengchiara) [reply kvm/v12.1.0/README.md:3]: @sslavic @snizhana-dynnyk I think the registry mirrors is a pretty significant change and we should call out in the main paragraph. Is this wrong? :D

Comment (cokiengchiara) [reply kvm/v12.1.0/README.md:3]: Should we use the language used in the AWS release? https://github.com/giantswarm/releases/tree/master/aws/v11.5.0

File: kvm/v12.1.0/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+## :zap:  Giant Swarm Release 12.1.0 for KVM :zap:
+
+**This release includes two significant improvements to NGINX Ingress Controller. It also adds support for image registry mirrors, improving reliability.**
+
+**The two NGINX Ingress Controller improvements:**
+* First, multiple NGINX Ingress Controllers per tenant cluster are now supported, enabling separation of internal vs external traffic, dev vs prod, and so on.
+* Second, management of NGINX IC NodePort Service is moved from kvm-operator to NGINX IC App itself. This enables configurability of external traffic policy and lays the foundation for making NGINX IC App optional and not pre-installed in a future KVM platform release.
+- NGINX IC NodePort Service external traffic policy has been made configurable.
```
Comment (cokiengchiara) [inline kvm/v12.1.0/README.md:8]:
```suggestion
```

File: kvm/v12.1.0/README.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+- Management of NGINX IC NodePort Service is moved from kvm-operator to NGINX IC App itself to:
```
Comment (cokiengchiara) [inline kvm/v12.1.0/README.md:9]:
```suggestion
```

File: kvm/v12.1.0/README.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  - Enable external traffic policy configurability, in a way consistent with other NGINX IC configuration options.
```
Comment (cokiengchiara) [inline kvm/v12.1.0/README.md:10]:
```suggestion
```

File: kvm/v12.1.0/README.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  - Lay the foundation for making NGINX IC App optional and not pre-installed in a future KVM platform release.
```
Comment (cokiengchiara) [inline kvm/v12.1.0/README.md:11]:
```suggestion
```

File: kvm/v12.1.0/README.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+**To change the ownership, `nginx-ingress-controller` NodePort Service in `kube-system` namespace needs to be manually deleted. It then gets to be recreated by triggering the cluster upgrade. To minimize the downtime, please delegate cluster upgrades to your SE.**
```
Comment (cokiengchiara) [inline kvm/v12.1.0/README.md:9]:
```suggestion
**Note for cluster upgrades:**

Please manually delete `nginx-ingress-controller` NodePort Service in `kube-system` namespace. Upgrading the cluster then recreates the NodePort Service and moves its management from kvm-operator to NGINX IC. To minimize downtime, please delegate cluster upgrades to your SE.**
```

@sslavic 
1. is this a required step? Or is it optional? 
1a If it is required (this is my impression), then I have suggestion in release notes. Pls check if makes it clearer.
1b If it's optional, then the instructions might be better placed in docs.
2. Also, this note applies to cluster upgrades only right? Would it be helpful to make that clearer?

File: kvm/v12.1.0/README.md (hunk 1 above)
Comment (snizhana-dynnyk) [reply kvm/v12.1.0/README.md]: yes, but without mentioning China

File: kvm/v12.1.0/README.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
+## :zap:  Giant Swarm Release 12.1.0 for KVM :zap:
+
+**This release includes two significant improvements to NGINX Ingress Controller. It also adds support for image registry mirrors, improving reliability.**
```
Comment (snizhana-dynnyk) [inline kvm/v12.1.0/README.md:3]:
```suggestion
**This release includes two significant improvements to NGINX Ingress Controller. It also includes a fix for Quay being a single point of failure by using Docker mirroring feature.
This ensures availability of all images needed for node bootstrap, thus the cluster creation / scaling doesn’t depend on Quay availability anymore.**
```

File: kvm/v12.1.0/release.yaml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    version: 2512.2.1
+  - name: etcd
+    version: 3.4.9
+  - name: flannel-operator
+    version: 0.2.0
+  - name: kubernetes
+    version: 1.17.8
+  - name: kvm-operator
+    version: 3.12.1
+  date: "2020-07-28T15:00:00Z"
```
Comment (sslavic) [inline kvm/v12.1.0/release.yaml:48]:
```suggestion
  date: "2020-07-29T15:00:00Z"
```

Comment (sslavic) [reply kvm/v12.1.0/release.yaml:48]: We'll release tomorrow

**Comment by sslavic** [comment]
Updated the release to include latest kube-state-metrics, metrics-server and node-exporter app releases.
Release notes have been updated accordingly.
I've retested on geckon:
- 12.1.0 cluster creation
- upgrade from 11.3.2

File: kvm/v12.1.0/README.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
+## :zap:  Giant Swarm Release 12.1.0 for KVM :zap:
+
+**This release includes two significant improvements to NGINX Ingress Controller. It also includes a fix for Quay being a single point of failure by using Docker mirroring feature. This ensures availability of all images needed for node bootstrap, thus the cluster creation/scaling doesn’t depend on Quay availability anymore.**
```
Comment (cokiengchiara) [inline kvm/v12.1.0/README.md:3]:
```suggestion
This release includes two **significant improvements to NGINX Ingress Controller**. It also includes a **fix for Quay being a single point of failure** by using Docker mirroring feature. This ensures availability of all images needed for node bootstrap, thus the cluster creation/scaling doesn’t depend on Quay availability anymore.
```

Formatting to bold only the important parts.

File: kvm/v12.1.0/README.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
+## :zap:  Giant Swarm Release 12.1.0 for KVM :zap:
+
+This release includes two **significant improvements to NGINX Ingress Controller**. It also includes a **fix for Quay being a single point of failure** by using Docker mirroring feature. This ensures availability of all images needed for node bootstrap, thus the cluster creation/scaling doesn’t depend on Quay availability anymore.
+
+**The two NGINX Ingress Controller improvements:**
+
+- First, multiple NGINX Ingress Controllers per tenant cluster are now supported, enabling separation of internal vs external traffic, dev vs prod, and so on.
```
Comment (cokiengchiara) [inline kvm/v12.1.0/README.md:7]:
```suggestion
- Multiple NGINX Ingress Controllers per tenant cluster are now supported, enabling separation of internal vs external traffic, dev vs prod, and so on.
```

File: kvm/v12.1.0/README.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+- Second, management of NGINX IC NodePort Service is moved from kvm-operator to NGINX IC App itself. This enables configurability of external traffic policy and lays the foundation for making NGINX IC App optional and not pre-installed in a future KVM platform release.
```
Comment (cokiengchiara) [inline kvm/v12.1.0/README.md:8]:
```suggestion
- Management of NGINX IC NodePort Service is moved from kvm-operator to NGINX IC App itself. This enables configurability of external traffic policy and lays the foundation for making NGINX IC App optional and not pre-installed in a future KVM platform release.
```

File: kvm/v12.1.0/announcement.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+This release includes two **significant improvements to NGINX Ingress Controller**. It also includes a **fix for Quay being a single point of failure** by using Docker mirroring feature.
+
+**The two NGINX Ingress Controller improvements:**
+
+- Multiple NGINX Ingress Controllers per tenant cluster are now supported
+- Management of NGINX IC NodePort Service is moved from kvm-operator to NGINX IC App itself
+
+Along with kvm-operator, cluster-operator, and NGINX IC, release includes other upstream component upgrades.
+
+You can find more details on components that were changed with this release: https://github.com/giantswarm/releases/tree/master/kvm/v12.1.0
```
Comment (cokiengchiara) [inline kvm/v12.1.0/announcement.md:12]:
```suggestion
You can find more details on components that were changed in [the release notes](https://github.com/giantswarm/releases/tree/master/kvm/v12.1.0).
```

**Comment by sslavic** [comment]
@cokiengchiara release has been made available to our customers. Thanks for all the help!

**Pull request merged**